crates/shared/src/price_estimation/trade_verifier/balance_overrides/detector.rs

```diff
@@ -39,121 +43,287 @@ impl std::ops::Deref for Detector {
-            let entry_points = [
-                // solc lays out memory linearly starting at 0 by default
-                "0000000000000000000000000000000000000000000000000000000000000000",
-            ];
-            for start_slot in entry_points {
-                let mut slot = U256::from(start_slot);
-                for _ in 0..probing_depth {
-                    strategies.push(Strategy::SolidityMapping { slot });
-                    slot += U256::one();
-                }
+    let mut buf = [0; 64];
+    buf[12..32].copy_from_slice(holder.as_slice());
+    for i in 0..heuristic_depth {
+        U256::from(i).to_big_endian(&mut buf[32..64]);
+        let slot_hash = B256::from(signing::keccak256(&buf));
+        heuristic_slot_to_index.insert(slot_hash, i);
+    }
+
+    buf[0..20].copy_from_slice(holder.as_slice());
+    buf[20..32].copy_from_slice(&[
+        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x87, 0xa2, 0x11, 0xa2,
+    ]);
+    let solady_slot = B256::from(signing::keccak256(&buf[0..32]));
+
+    // Create strategies for each slot
+    let mut strategies: Vec<(Strategy, bool)> = storage_slots
+        .iter()
+        .map(|(contract, slot)| {
+            if let Some(&map_slot_index) = heuristic_slot_to_index.get(slot) {
+                // This slot matches a SolidityMapping heuristic
+                (
+                    Strategy::SolidityMapping {
+                        target_contract: contract.into_legacy(),
+                        map_slot: U256::from(map_slot_index),
+                    },
+                    true, // is_heuristic
+                )
+            } else if *slot == solady_slot {
+                // This slot matches a SoladyMapping heuristic
+                (
+                    Strategy::SoladyMapping {
+                        target_contract: contract.into_legacy(),
+                    },
+                    true, // is_heuristic
+                )
+            } else {
+                // This slot doesn't match heuristic, use DirectSlot
+                (
+                    Strategy::DirectSlot {
+                        target_contract: contract.into_legacy(),
+                        slot: slot.into_legacy(),
+                    },
+                    false, // not heuristic
+                )
             }
+        })
+        .collect();
 
-            strategies
-                .into_iter()
-                .enumerate()
-                .map(|(index, strategy)| StrategyHelper::new(strategy, index))
-                .collect()
-        };
+    // Sort by whether it's a heuristic slot (stable sort preserves order)
+    strategies.sort_by_key(|(_, is_heuristic)| *is_heuristic);
 
-        let state_overrides = strategies
-            .iter()
-            .map(|helper| helper.strategy.state_override(&holder, &helper.balance))
-            .collect::<HashMap<_, _>>();
+    // Reverse so most recently accessed come first
+    strategies.reverse();
```

Comment:
It can probably be done in one pass.
```suggestion
    // Sort by whether it's a heuristic slot (stable sort preserves order)
    strategies.sort_by_key(|(_, is_heuristic)| {
        // Reverse so most recently accessed come first
        (!*is_heuristic,)
    });
```